processing/src/main/java/org/apache/druid/segment/ReferenceCountingSegment.java

```diff
@@ -79,6 +80,9 @@ protected ReferenceCountingSegment(
   )
   {
     super(baseSegment);
+    if (baseSegment instanceof SegmentReference) {
+      throw DruidException.defensive("Cannot use a SegmentReference as baseSegment for a ReferenceCountingSegment");
```

Comment:
Might be nice to include segment asString() output in exception message so we know what kind of segment we tried to accidentally wrap as a reference counter

```suggestion
      throw DruidException.defensive("Cannot use a SegmentReference[%s] as baseSegment for a ReferenceCountingSegment", baseSegment.asString());
```

Also i think maybe this constructor could be private?

Reply:
updated this